conda-store-server/tests/test_db_api.py

```diff
@@ -39,35 +39,104 @@ def test_namespace_crud(db):
+    # Attempts to create a role mapping with an invalid role
+    with pytest.raises(ValueError, match=r"invalid role=invalid-role"):
+        api.create_namespace_role(db, name=namespace_name, other=other_namespace_name3, role="invalid-role")
+        db.commit()
+
+    # Attempts to create a role mapping violating the uniqueness constraint
+    with pytest.raises(
+            Exception,
+            match=(r"UNIQUE constraint failed: "
+                   r"namespace_role_mapping.namespace_id, "
+                   r"namespace_role_mapping.other_namespace_id")):
+        # Runs in a nested transaction since a constraint violation will cause a rollback
+        with db.begin_nested():
+            api.create_namespace_role(db, name=namespace_name, other=other_namespace_name2, role="developer")
+            db.commit()
+
+    # Updates a role mapping
+    api.update_namespace_role(db, name=namespace_name, other=other_namespace_name2, role="developer")
+    db.commit()
 
-    # Creates role mappings with valid entity names
-    NamespaceRoleMapping(namespace=namespace, namespace_id=namespace.id, entity="org/*")
+    # Gets all role mappings
+    roles = api.get_namespace_roles(db, namespace_name)
+    db.commit()
+    assert len(roles) == 3
 
-    NamespaceRoleMapping(
-        namespace=namespace, namespace_id=namespace.id, entity="*/team"
-    )
+    assert roles[0].id == 1
+    assert roles[0].namespace == namespace_name
+    assert roles[0].other_namespace == other_namespace_name1
+    assert roles[0].role == 'admin'
 
-    NamespaceRoleMapping(
-        namespace=namespace, namespace_id=namespace.id, entity="org/team"
-    )
+    assert roles[1].id == 2
+    assert roles[1].namespace == namespace_name
+    assert roles[1].other_namespace == other_namespace_name2
+    assert roles[1].role == 'developer'
+
+    assert roles[2].id == 3
+    assert roles[2].namespace == namespace_name
+    assert roles[2].other_namespace == other_namespace_name3
+    assert roles[2].role == 'viewer'
 
-    NamespaceRoleMapping(namespace=namespace, namespace_id=namespace.id, entity="*/*")
+    # Gets other role mappings
+    roles = api.get_other_namespace_roles(db, other_namespace_name1)
+    db.commit()
+    assert len(roles) == 1
+    roles = api.get_other_namespace_roles(db, namespace_name)
+    db.commit()
+    assert len(roles) == 0
+
+    # Deletes one role mapping
+    api.delete_namespace_role(db, name=namespace_name, other=other_namespace_name2)
+    db.commit()
+
+    # Gets all role mappings again
+    roles = api.get_namespace_roles(db, namespace_name)
+    db.commit()
+    assert len(roles) == 2
+
+    assert roles[0].id == 1
+    assert roles[0].namespace == namespace_name
+    assert roles[0].other_namespace == other_namespace_name1
+    assert roles[0].role == 'admin'
+
+    assert roles[1].id == 3
+    assert roles[1].namespace == namespace_name
+    assert roles[1].other_namespace == other_namespace_name3
+    assert roles[1].role == 'viewer'
```

Comment:
Delete one mapping and see how it's reflected in the DB. Namespace2 is deleted as it should.